libraries/api/application/ApiApplication.php

```diff
@@ -47,71 +56,118 @@ public function __construct(\JInput $input = null, Registry $config = null, WebC
 
+		// Setup the router
+		// TODO: Router class not ready
+		// $this->router = new ApiRouter();
 	}
 
+
 	/**
-	 * Dispatch the application
+	 * Method to run the application routines.
 	 *
-	 * @param   string $component The component which is being rendered.
+	 * Most likely you will want to instantiate a controller and execute it, or perform some sort of task directly.
 	 *
 	 * @return  void
 	 *
-	 * @since   3.2
+	 * @since   1.0
 	 */
-	public function dispatch($component = null)
+	protected function doExecute()
 	{
+		// Initialise the application
+		$this->initialiseApp();
 
 	}
 
 	/**
-	 * Method to run the application routines.
-	 *
-	 * Most likely you will want to instantiate a controller and execute it, or perform some sort of task directly.
+	 * Rendering is the process of pushing the document buffers into the template
+	 * placeholders, retrieving data from the document and pushing it into
+	 * the application response buffer.
 	 *
 	 * @return  void
 	 *
-	 * @since   1.0
+	 * @since   3.2
+	 *
+	 * @note    Rendering should be overridden to get rid of the theme files.
 	 */
-	protected function doExecute()
+	protected function render()
 	{
-
+		// Render the document
+		$this->setBody($this->document->render($this->allowCache()));
 	}
 
 	/**
-	 * Initialise the application.
-	 *
-	 * @param   array $options An optional associative array of configuration settings.
+	 * Method to send the application response to the client.  All headers will be sent prior to the main application output data.
 	 *
 	 * @return  void
 	 *
-	 * @since   3.2
+	 * @since   1.0
 	 */
-	protected function initialiseApp($options = array())
+	protected function respond()
 	{
-
+		// Send the content-type header.
+		if (!$this->getResponse()->hasHeader('Content-Type'))
+		{
+			$this->setHeader('Content-Type', $this->mimeType . '; charset=' . $this->charSet);
+		}
+
+		// If the response is set to uncachable, we need to set some appropriate headers so browsers don't cache the response.
+		if (!$this->allowCache())
+		{
+			// Expires in the past.
+			$this->setHeader('Expires', 'Wed, 17 Aug 2005 00:00:00 GMT', true);
+
+			// Always modified.
+			$this->setHeader('Last-Modified', gmdate('D, d M Y H:i:s') . ' GMT', true);
+			$this->setHeader('Cache-Control', 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0', false);
+
+			// HTTP 1.0
+			$this->setHeader('Pragma', 'no-cache');
+		}
+		else
+		{
+			// Expires.
+			if (!$this->getResponse()->hasHeader('Expires'))
+			{
+				$this->setHeader('Expires', gmdate('D, d M Y H:i:s', time() + 900) . ' GMT');
+			}
+
+			// Last modified.
+			if (!$this->getResponse()->hasHeader('Last-Modified') && $this->modifiedDate instanceof \DateTime)
+			{
+				$this->modifiedDate->setTimezone(new \DateTimeZone('UTC'));
+				$this->setHeader('Last-Modified', $this->modifiedDate->format('D, d M Y H:i:s') . ' GMT');
+			}
+		}
+
+		$this->sendHeaders();
+
+		echo json_encode($this->getBody());
```

Comment:
I mean part of it depends whether we will ever ever support non-json. There's still a bit of me that thinks JDocument::render should just be a string (consistency with JDocumentHtml) and that is doing the json_encode, which gives us freedom to do XML or alternative doc types in the future. I mean the actual work on the non-json types are kinda out of scope for this GSOC but neither should be outright discounted because we only json_encode in the app